lib/src/gitattributes.rs

```diff
@@ -0,0 +1,705 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![expect(missing_docs)]
+#![allow(unused)]
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::fs::{self};
+use std::io::Cursor;
+use std::io::ErrorKind;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+use gix_attributes::Search;
+use gix_attributes::State;
+use gix_attributes::glob::pattern::Case;
+use gix_attributes::search::MetadataCollection;
+use gix_attributes::search::Outcome;
+use pollster::FutureExt as _;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncReadExt as _;
+use tokio::sync::OnceCell;
+
+use crate::backend::TreeValue;
+use crate::file_util::BlockingAsyncReader;
+use crate::merge::SameChange;
+use crate::merged_tree::MergedTree;
+use crate::repo_path::RepoPath;
+use crate::repo_path::RepoPathBuf;
+use crate::repo_path::RepoPathComponent;
+
+/// Git Attributes instance
+///
+/// This struct handles accessing .gitattributes files
+/// and maintains a cache to prevent loading the same file twice.
+///
+/// It's lazy loaded so .gitattributes files are only accessed
+/// if you search for attributes for a given path.
+pub(crate) struct GitAttributes {
+    disk_file_loader: Arc<dyn FileLoader>,
+    store_file_loader: Arc<dyn FileLoader>,
+    node_cache: Mutex<HashMap<RepoPathBuf, Arc<GitAttributesNode>>>,
+}
+
+impl std::fmt::Debug for GitAttributes {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GitAttributes").finish()
+    }
+}
+
+#[async_trait::async_trait]
+pub(crate) trait FileLoader: Send + Sync {
+    /// Loads a file in a given `path`
+    ///
+    /// Returns Some(..) if the file was found and None if not
+    async fn load(
+        &self,
+        path: &RepoPath,
+    ) -> Result<Option<Box<dyn AsyncRead + Send + Unpin>>, GitAttributesError>;
+}
+
+pub(crate) struct TreeFileLoader {
+    tree: MergedTree,
+}
+impl TreeFileLoader {
+    pub fn new(tree: MergedTree) -> Self {
+        Self { tree }
+    }
+}
+
+#[async_trait::async_trait]
+impl FileLoader for TreeFileLoader {
+    async fn load(
+        &self,
+        path: &RepoPath,
+    ) -> Result<Option<Box<dyn AsyncRead + Send + Unpin>>, GitAttributesError> {
+        let merged_tree_value =
+            self.tree
+                .path_value_async(path)
+                .await
+                .map_err(|err| GitAttributesError {
+                    message: "Could not retrieve the value from path".to_string(),
+                    source: err.into(),
+                })?;
+        let Some(file_to_merge) = merged_tree_value.to_file_merge() else {
+            // this means the file doesn't exist in the tree
```

Comment:
This assumption is wrong: if the merge has multiple terms, and *one of* the terms is not a file(e.g. one term is a folder, another term is a file; a more practical case will be one term is a symlink, another term is a file), `None` will also be returned. A better way to handle this case is to fallback to check if one of the terms is file, and use that. Check https://github.com/jj-vcs/jj/compare/main...06393993:jj:gitattr-example-impl#diff-09e46fa0436a80b91b08a5861cd4df4ee1a87865c4853a055d78eb37a334202aR74-R92 on how to implement such fallback.

Having said so, I don't think we have to implement such fallback, we can do it in a follow-up PR with tests.

Reply:
Ok, using the same code as the example.